src/ic-cdk-codegen/src/lib.rs

```diff
@@ -0,0 +1,553 @@
+use candid::Principal;
+use candid::parser::types::Dec;
+use num_bigint::BigUint;
+use proc_macro2::Span;
+use syn::*;
+use proc_macro2::TokenStream;
+use candid::parser::types::{IDLType, PrimType, Binding};
+use candid::types::internal::Label;
+use syn::spanned::Spanned;
+use std::result::Result as StdResult;
+use std::error::Error as StdError;
+use std::fs;
+use std::path::Path;
+
+
+/// Processes Candid declarations into Rust symbols.
+#[derive(Debug, Clone)]
+pub struct Processor {
+    decls: HashMap<String, TokenStream>,
+    bindings: HashMap<String, Signature>,
+    record_id: usize,
+    variant_id: usize,
+}
+
+impl Default for Processor {
+    fn default() -> Self {
+        Self { 
+            decls: initial_prim_map(),
+            bindings: HashMap::new(),
+            record_id: 0,
+            variant_id: 0,
+        }
+    }
+}
+
+impl Processor {
+    /// Constructs a `Processor` from an existing set of symbols. 
+    /// 
+    /// These symbols will override those in the Candid file, as well as built-in types like `Nat`.
+    /// Buit-in type names draw from [`PrimType`], not the Candid language.
+    /// 
+    /// # Rules
+    /// 
+    /// - Symbols must be `struct`s, `enum`s, `type`s, `use`s, or `async fn`s.
+    /// - All symbols must be `pub`.
+    /// - `use`s must not be globs (`*`).
+    /// - Nothing may be generic.
+    /// - Functions may neither be `const` nor `extern`.
+    /// - Functions should, but do not have to, end in `;` instead of `{}`.
+    pub fn new(body: Vec<Item>) -> Result<Self> {
+        let (predecls, bindings) = body_to_decls(body)?;
+        let mut decls = initial_prim_map();
+        decls.extend(predecls);
+        Ok(Self { decls, bindings, record_id: 0, variant_id: 0 })
+    }
+    
+    fn next_record(&mut self) -> usize {
+        let ret = self.record_id;
+        self.record_id += 1;
+        ret
+    }
+
+    fn next_variant(&mut self) -> usize {
+        let ret = self.variant_id;
+        self.variant_id += 1;
+        ret
+    }
+
+    /// Adds a [`Binding`] to the type declarations. 
+    /// 
+    /// These types will be attempted to be exported from the generated code, 
+    /// but in the case of name conflicts, actor functions will override them.
+    /// In this case the type will be accessible from a module named `t`.
+    pub fn add_decl(&mut self, decl: Binding) -> Result<()> {
+        self.idl_to_rust(&decl.typ, Some(&decl.id))?;
+        Ok(())
+    }
+
+    fn idl_to_rust(&mut self, ty: &IDLType, name: Option<&str>) -> Result<TokenStream> {
+        if let Some(name) = name {
+            if self.decls.contains_key(name) {
+                let (ident, _) = str_to_ident(name);
+                return Ok(quote!(#ident));
+            }
+        }
+        let ret = match ty {
+            IDLType::RecordT(fields) => {
+                let name = name
+                    .map(String::from)
+                    .unwrap_or_else(|| format!("_R{}", self.next_record()));
```

Comment:
How does user know the struct name and use it in their code?

Reply:
The user would use their editor's autocompletion to see what types the function uses. There really isn't a better way as far as I can tell to do that - if the user's editor can't autocomplete into proc-macro generated code, that is what the build.rs version is good for.

Reply:
Autocompletion is not always available. And all the ids can suddenly change if you add a new line in the did file.

Reply:
What naming scheme would you suggest?

Reply:
Extending the interface will also cause the id to change. I don't think there is a good way to generate names unless a user is involved. Proc macro generated code is hidden from the user, which makes it difficult to work with. Generating a binding from `didc`, and make that binding part of the source code is a good direction to go. The proc macro would then just check for subtype to make ensure the binding to consistent with the did file.

Reply:
I think there are two aspects of codegen being mixed here - that it autogenerates names, and that it's in a proc macro. Whatever solution there is for names can be implemented whether it's a proc macro or a standalone code generator, and vice versa. That's why I was told to make it a separate codegen crate instead of putting the entire thing in ic_cdk_macros - this is capable of generating binding files too. Adding members to the macro-annotated module declaration and modifying the generated code in a source file are two ways of doing the same thing. I'm just trying to get it so that, regardless of which way you use the codegen, you *don't* have to write all your own declarations when the candid file contains inline things.